src/host-config.ts

```diff
@@ -164,7 +178,36 @@ async function readHostname() {
 
 async function setHostname(val: string) {
 	await config.set({ hostname: val });
-	await dbus.restartService('resin-hostname');
+	if (await isServiceLoaded('balena-hostname')) {
+		if (!(await isServicePartOf('balena-hostname', 'config-json.target'))) {
+			await dbus.restartService('balena-hostname');
+		}
+	} else if (await isServiceLoaded('resin-hostname')) {
+		await dbus.restartService('resin-hostname');
+	}
+}
+
+async function isServiceLoaded(serviceName: string): Promise<boolean> {
+	try {
+		return (await dbus.serviceLoadState(serviceName)) === 'loaded';
+	} catch (e) {
+		return false;
+	}
+}
+
+async function isServicePartOf(
+	serviceName: string,
+	unitName: string,
+): Promise<boolean> {
+	try {
+		const partOf = await dbus.servicePartOf(serviceName);
+		return _.includes(unitName, partOf);
```

Comment:
At this point `partOf` is an unknown type and I can't do much with it. The following line always fails, even when I print `partOf` and confirm the substring exists.

Reply:
think the problem is that `getUnitProperty` doesn't have a return type. Can you add `Promise<string>` as return type to the function and try